extensions/amp-subscriptions/0.1/local-subscription-platform-base.js

```diff
@@ -0,0 +1,192 @@
+    );
+
+    /** @private @const {!LocalSubscriptionPlatformRenderer}*/
+    this.renderer_ = new LocalSubscriptionPlatformRenderer(this.ampdoc_,
+        serviceAdapter.getDialog(), this.serviceAdapter_);
+  }
+
+  /**
+   * @override
+   */
+  getServiceId() {
+    return 'local';
+  }
+
+  /**
+   * Validates the action map
+   * @param {!JsonObject<string, string>} actionMap
+   * @return {!JsonObject<string, string>}
+   */
+  validateActionMap(actionMap) {
+    userAssert(actionMap['login'],
+        'Action "login" is not present in action map');
+    userAssert(actionMap['subscribe'],
+        'Action "subscribe" is not present in action map');
+    return actionMap;
+  }
+
+  /**
+   * Add event listener for the subscriptions action
+   * @protected
+   */
+  initializeListeners_() {
+    this.rootNode_.addEventListener('click', e => {
+      const element =
+        closestAncestorElementBySelector(dev().assertElement(e.target),
+            '[subscriptions-action]');
+      this.handleClick_(element);
+    });
+  }
+
+  /**
+   * Handle click on subscription-action
+   * @private
+   * @param {Node} element
+   */
+  handleClick_(element) {
+    if (element) {
+      const action = element.getAttribute('subscriptions-action');
+      const serviceAttr = element.getAttribute('subscriptions-service');
+      if (serviceAttr == 'local') {
+        this.executeAction(action);
+      } else if ((serviceAttr || 'auto') == 'auto') {
+        if (action == 'login') {
+          // The "login" action is somewhat special b/c viewers can
+          // enhance this action, e.g. to provide save/link feature.
+          const platform = this.serviceAdapter_.selectPlatformForLogin();
+          this.serviceAdapter_.delegateActionToService(
+              action, platform.getServiceId());
+        } else {
+          this.executeAction(action);
+        }
+      } else if (serviceAttr) {
+        this.serviceAdapter_.delegateActionToService(action, serviceAttr);
+      }
+    }
+  }
+
+  /** @override */
+  activate(entitlement) {
+    const renderState = entitlement.json();
+    this.urlBuilder_.setAuthResponse(renderState);
+    this.actions_.build().then(() => {
+      this.renderer_.render(renderState);
+    });
+  }
+
+  /** @override */
+  reset() {
+    this.renderer_.reset();
+  }
+
+  /** @override */
+  executeAction(action) {
+    const actionExecution = this.actions_.execute(action);
+    return actionExecution.then(result => {
+      if (result) {
+        this.serviceAdapter_.resetPlatforms();
+      }
+      return !!result;
+    });
+  }
+
+  /** @override */
+  isPrerenderSafe() {
+    // Local platform can never be allowed to prerender in a viewer
+    return false;
+  }
+
+  /** @override */
+  getSupportedScoreFactor(unusedFactor) {
+    return 0;
+  }
+
+  /** @override */
+  getBaseScore() {
+    return this.serviceConfig_['baseScore'] || 0;
+  }
+
+  /** @override */
+  getEntitlements() {
```

Comment:
Should we make `getEntitlements` and `pingback` abstract?

Reply:
done